stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -490,10 +497,106 @@ impl NakamotoBlockHeader {
+    pub fn verify_signer_signatures(&self, reward_set: &RewardSet) -> Result<(), ChainstateError> {
+        let message = self.signer_signature_hash();
+        let Some(signers) = &reward_set.signers else {
+            return Err(ChainstateError::InvalidStacksBlock(
+                "No signers in the reward set".to_string(),
+            ));
+        };
+
+        let mut total_weight_signed: u32 = 0;
+        // `last_index` is used to prevent out-of-order signatures
+        let mut last_index = None;
+
+        let total_weight = signers.iter().map(|s| s.weight).fold(0, |w, acc| {
+            acc.checked_add(w)
+                .expect("FATAL: Total signer weight > u32::MAX")
+        });
+
+        // HashMap of <PublicKey, (Signer, Index)>
+        let signers_by_pk = signers
+            .iter()
+            .enumerate()
+            .map(|(i, signer)| (&signer.signing_key, (signer, i)))
+            .collect::<HashMap<_, _>>();
+
+        for signature in self.signer_signature.iter() {
+            let public_key = Secp256k1PublicKey::recover_to_pubkey(message.bits(), signature)
+                .map_err(|_| {
+                    ChainstateError::InvalidStacksBlock(format!(
+                        "Unable to recover public key from signature {}",
+                        signature.to_hex()
+                    ))
+                })?;
+
+            let mut public_key_bytes = [0u8; 33];
+            public_key_bytes.copy_from_slice(&public_key.to_bytes_compressed()[..]);
+
+            let (signer, signer_index) = signers_by_pk.get(&public_key_bytes).ok_or_else(|| {
+                ChainstateError::InvalidStacksBlock(format!(
+                    "Public key {} not found in the reward set",
+                    public_key.to_hex()
+                ))
+            })?;
+
+            // Enforce order of signatures
+            if let Some(index) = last_index.as_ref() {
+                if *index >= *signer_index {
+                    return Err(ChainstateError::InvalidStacksBlock(
+                        "Signatures are out of order".to_string(),
+                    ));
+                }
+            } else {
+                last_index = Some(*signer_index);
+            }
+
+            total_weight_signed = total_weight_signed
+                .checked_add(signer.weight)
+                .expect("FATAL: overflow while computing signer set threshold");
+        }
+
+        let threshold = Self::compute_voting_weight_threshold(total_weight)?;
+
+        if total_weight_signed < threshold {
+            return Err(ChainstateError::InvalidStacksBlock(format!(
+                "Not enough signatures. Needed at least {} but got {}",
+                threshold, total_weight_signed
+            )));
+        }
+
+        return Ok(());
+    }
+
+    pub fn compute_voting_weight_threshold(total_weight: u32) -> Result<u32, ChainstateError> {
+        let ceil = if (total_weight as u64 * 7) % 10 == 0 {
```

Comment:
We try to avoid the `as` typecast since it can implicitly truncate data.  Instead, please do this:

```rust
let ceil = if (u64::from(total_weight) * 7) % 10 == 0 {
```